### PR TITLE
add section for host header attack prevention in rails security guide

### LINKED ISSUE #42817: Security guide doesn’t cover host authorization/DNS rebinding protection
Action Dispatch comes with host authorization, a security feature that can be enabled in production by populating `config.hosts`. The DNS rebinding attacks it protects against aren’t covered in [the security guide](https://edgeguides.rubyonrails.org/security.html). The guide should cover the relevant security issues, how to protect against them at the app level with `config.hosts`, and how reverse proxies and load balancers like Nginx sometimes protect against them already.

### PR DESCRIPTION
**Pull request by aka47**
This PR fixes rails/rails#42817

It is also fixes a related issue in the rails configuration guide about the HostAuthorization middleware - https://edgeguides.rubyonrails.org/configuring.html#actiondispatch-hostauthorization

### CODE REVIEW

**Comment by p8** [comment]
This looks great, thanks!
Also great find with that incorrect `host_configuration` in `configuring.md`.

**Review by p8 — CHANGES_REQUESTED** [review]

File: guides/source/security.md [hunk 1]
```diff
@@ -949,6 +949,27 @@ Location: http://www.malicious.tld
...
 
+#### DNS rebinding and Host header attacks
+
+DNS rebinding is a method of manipulating resolution of domain names that is commonly used as a form of computer attack. DNS rebinding circumvents the same-origin policy by abusing the Domain Name System (DNS) instead. It rebinds a domain to a different IP address and than compromises the system by executing random code against your Rails app from the changed IP address.
+
+It is recommended to use the ActionDispatch::HostAuthorization middleware to guard against DNS rebinding and other Host header attacks. It is enabled by default in the development environment, you have to activate it in production and other environments by setting the list of allowed hosts. You can also configure exceptions and set your own response app.
```
Comment (rafaelfranca) [inline guides/source/security.md:956]:
```suggestion
It is recommended to use the `ActionDispatch::HostAuthorization` middleware to guard against DNS rebinding and other Host header attacks. It is enabled by default in the development environment, you have to activate it in production and other environments by setting the list of allowed hosts. You can also configure exceptions and set your own response app.
```

File: guides/source/security.md [hunk 2]
```diff
@@ -949,6 +949,27 @@ Location: http://www.malicious.tld
...
+  # Exclude requests for the /healthcheck/ path from host checking
+  exclude: ->(request) { request.path =~ /healthcheck/ }
+  # Add custom Rack application for the response
+  response_app: -> env do
+    [400, { "Content-Type" => "text/plain" }, ["Bad Request"]]
+  end
+}
+```
+
+You can read more about it in the [ActionDispatch::HostAuthorization middleware documentation](/configuring.html#actiondispatch-hostauthorization)
```
Comment (rafaelfranca) [inline guides/source/security.md:971]:
```suggestion
You can read more about it in the [`ActionDispatch::HostAuthorization` middleware documentation](/configuring.html#actiondispatch-hostauthorization)
```

File: guides/source/security.md [hunk 3]
```diff
@@ -949,6 +949,27 @@ Location: http://www.malicious.tld
 
 So _attack vectors for Header Injection are based on the injection of CRLF characters in a header field._ And what could an attacker do with a false redirection? They could redirect to a phishing site that looks the same as yours, but ask to login again (and sends the login credentials to the attacker). Or they could install malicious software through browser security holes on that site. Rails 2.1.2 escapes these characters for the Location field in the `redirect_to` method. _Make sure you do it yourself when you build other header fields with user input._
 
+#### DNS rebinding and Host header attacks
```
Comment (p8) [inline guides/source/security.md:952]: Most words are capitalized in headings:
https://guides.rubyonrails.org/ruby_on_rails_guides_guidelines.html#headings
```suggestion
#### DNS Rebinding and Host Header Attacks
```

File: guides/source/security.md [hunk 4]
```diff
@@ -949,6 +949,27 @@ Location: http://www.malicious.tld
...
+
+DNS rebinding is a method of manipulating resolution of domain names that is commonly used as a form of computer attack. DNS rebinding circumvents the same-origin policy by abusing the Domain Name System (DNS) instead. It rebinds a domain to a different IP address and than compromises the system by executing random code against your Rails app from the changed IP address.
+
+It is recommended to use the `ActionDispatch::HostAuthorization` middleware to guard against DNS rebinding and other Host header attacks. It is enabled by default in the development environment, you have to activate it in production and other environments by setting the list of allowed hosts. You can also configure exceptions and set your own response app.
```
Comment (p8) [inline guides/source/security.md:956]: As this middleware is optional on production (as most will be running Rails behind a proxy), I think we can reword it a little to make it less prescriptive. Maybe something like:
```suggestion
It is recommended to use the `ActionDispatch::HostAuthorization` middleware to guard against DNS rebinding and other Host header attacks. It is enabled by default in the development environment. To enable it in production and other environments set the list of allowed hosts. You can also configure exceptions and set your own response app.
```

**Review by p8 — CHANGES_REQUESTED** [review]
Sorry this has taken so long @aka47 😞 
This looks great.
Just some small changes, and this is good to go.

**Comment by p8** [comment]
Thanks @aka47 ! 🙌

**Pull request merged**